### PR TITLE
openacr: Nov 22 / 2023 upstream update

### PR DESCRIPTION
**Pull request by alexeilebedev**
* algo_lib:
- After ReadArgv, algo_lib::_db.errtext is non-empty
- merge algo_lib headers into include/algo.h, include/algo.inl.h

* amc:
- cfmt with anonfld must still print the anonfld name
- Recursively insert tuples into all dependent namespaces
- In ReadArgv, simply add LoadTuplesMaybe for all dependencies
- move LoadTuplesFd, LoadTuplesFile, DoLoadTuples to amc
- move recursion from ReadArgv to ns::LoadTuplesMaybe and ns::LoadTuplesFd

* abt_md:
- add -dry_run option
- add -print option
- add section selector
- detect unbalanced backticks in readme
- detect errors in sandbox command output (now you can process or print a single section)
- fail if any `inline-command` fails

* gcache: Implement as c++ binary
- implement -install, -disable, -enable, -stats, -clean, -gc
- use copy_file_range system call (30% faster)
- implement -hitrate -after <timestamp>
- clean log every 2 days
- prevent cached files used within last 2 days from being deleted
- feed preprocessed file to the compiler (faster)

* abt
- report hit rate after execution
- use temporary file for lib/exe out file,

* acr:
- ignore duplicate records by default This new behavior matches the behavior of acr where it printed the warning set the exit code to 1, but then still saved files to disk. This seems to be what people expect.
- proper implementation of -trunc
- automatically cascade-delete any remaining records after running
- load additional tables needed to pass `-check`
- implement -check -del
- add tests for delete&insert, insert&delete, trunc
- update documentation
- add -selerr flag (with -check)

* acr_ed: automatically create readme for new target
- show inserted tuples by default; use `-replace` by default; split `-acr -check` into a separate step
- -del -write -target ...: remove readme as well
- each target should have its files under cpp/$target
- rewrite Create_Target and Rename_Target to follow the uniform directory structure
- add a list of actions to dev.edacttype and dev.edaction
- warn if zero or more than one action is attempted during a run
- update documentation
- support -rename with -target
- support -alias option
- pick Pkey as reftype if a relation is being created between two ssimdbs

* acr_dm:
- place new tuples correctly within the merged file

* atf_comp:
- with -e, show exit_code, comment
- show filter on a separate line
- do not print '# atf_comp ...' line (for binary tests)

* samp_regx:
- sample for displaying innards of regx

* atf_ci:
- add an acr_ed for renaming a target

* src_func:
- add support for default arguments to avoid breeding functions
- fix target selection with regx fixes spurious compilation error when gcache/abt gets rebuilt along with everything else

* misc:
- unlink amc,acr,sv2ssim from lib_prot
- add lib_amcdb (for amcdb namespace)
- remove nossimfile entries
- rename tutorial samples so they don't overwrite handwritten file
- update documentation
- bootstrap: reduce size of bootstrap files, move _gen.cpp into algo_lib
- readme: split ns column into separate table

### CODE REVIEW

File: bin/atf_wcli [hunk 1]
```diff
@@ -0,0 +1 @@
+../build/release/atf_wcli
```
Comment (ali5h) [inline bin/atf_wcli:1]: rm

File: cpp/abt/opt.cpp [hunk 2]
```diff
@@ -39,6 +39,10 @@ tempstr abt::EvalSrcfileCmdline(algo_lib::Replscope &R, abt::FTarget &target, ab
     } else {
         if (abt::_db.gcache) {
             out << "gcache ";
```
Comment (ali5h) [inline cpp/abt/opt.cpp:41]: why not use command::gcache?

**Review by ali5h — APPROVED** [review]

File: bin/atf_wcli (hunk 1 above)
Comment (alexeilebedev) [reply bin/atf_wcli]: done.

File: cpp/abt/opt.cpp (hunk 2 above)
Comment (alexeilebedev) [reply cpp/abt/opt.cpp:41]: there are three variations -- gcache,cache, and no prefix. leaving it in prefix form is more convenient given that the bulk of the command line occurs after "--"

**Pull request merged**